velox/dwio/dwrf/common/FileMetadata.cpp

```diff
@@ -0,0 +1,131 @@
+ * limitations under the License.
+ */
+#include "velox/dwio/dwrf/common/FileMetadata.h"
+
+namespace facebook::velox::dwrf {
+namespace detail {
+using dwio::common::CompressionKind;
+
+CompressionKind orcCompressionToCompressionKind(
+    proto::orc::CompressionKind compression) {
+  switch (compression) {
+    case proto::orc::CompressionKind::NONE:
+      return CompressionKind::CompressionKind_NONE;
+    case proto::orc::CompressionKind::ZLIB:
+      return CompressionKind::CompressionKind_ZLIB;
+    case proto::orc::CompressionKind::SNAPPY:
+      return CompressionKind::CompressionKind_SNAPPY;
+    case proto::orc::CompressionKind::LZO:
+      return CompressionKind::CompressionKind_LZO;
+    case proto::orc::CompressionKind::LZ4:
+      return CompressionKind::CompressionKind_ZSTD;
+    case proto::orc::CompressionKind::ZSTD:
+      return CompressionKind::CompressionKind_LZ4;
+  }
+  return CompressionKind::CompressionKind_NONE;
+}
+} // namespace detail
+
+PostScript::PostScript(const proto::PostScript& ps)
+    : footerLength_{ps.footerlength()},
+      compression_{
+          ps.has_compression()
+              ? static_cast<dwio::common::CompressionKind>(ps.compression())
+              : dwio::common::CompressionKind::CompressionKind_NONE},
+      compressionBlockSize_{
+          ps.has_compressionblocksize()
+              ? ps.compressionblocksize()
+              : dwio::common::DEFAULT_COMPRESSION_BLOCK_SIZE},
+      writerVersion_{
+          ps.has_writerversion()
+              ? static_cast<WriterVersion>(ps.writerversion())
+              : WriterVersion::ORIGINAL},
+      cacheMode_{static_cast<StripeCacheMode>(ps.cachemode())},
+      cacheSize_{ps.cachesize()} {}
+
+PostScript::PostScript(const proto::orc::PostScript& ps)
+    : fileFormat_{dwio::common::FileFormat::ORC},
+      footerLength_{ps.footerlength()},
+      compression_{detail::orcCompressionToCompressionKind(ps.compression())},
+      compressionBlockSize_{ps.compressionblocksize()},
+      writerVersion_{static_cast<WriterVersion>(ps.writerversion())},
+      metadataLength_{ps.metadatalength()},
+      stripeStatisticsLength_{ps.stripestatisticslength()} {}
+
+TypeKind ProtoType::kind() const {
+  if (format_ == DwrfFormat::kDwrf) {
+    switch (dwrfPtr()->kind()) {
+      case proto::Type_Kind_BOOLEAN:
+      case proto::Type_Kind_BYTE:
+      case proto::Type_Kind_SHORT:
+      case proto::Type_Kind_INT:
+      case proto::Type_Kind_LONG:
+      case proto::Type_Kind_FLOAT:
+      case proto::Type_Kind_DOUBLE:
+      case proto::Type_Kind_STRING:
+      case proto::Type_Kind_BINARY:
+      case proto::Type_Kind_TIMESTAMP:
+        return static_cast<TypeKind>(dwrfPtr()->kind());
+      case proto::Type_Kind_LIST:
+        return TypeKind::ARRAY;
+      case proto::Type_Kind_MAP:
+        return TypeKind::MAP;
+      case proto::Type_Kind_UNION: {
+        DWIO_RAISE("Union type is deprecated!");
```

Comment:
UNION seems still supported in old code, you might need to find something else than `TypeKind` to represent the type to make sure all types currently supported are covered without throwing.

Reply:
Could you elaborate?
In https://github.com/facebookincubator/velox/blob/ff331272ad5f67ad5585085573c3401528282de1/velox/dwio/dwrf/reader/ReaderBase.cpp#L296 
`UNION` is shown deprecated, and there is no UNION in TypeKind. If we don't throw here, it will eventually throw at `convertType`.

Reply:
OK I missed that part.